File: backend/dataset/coinChange/coinChange_9.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        maxi = 100001
        dp = [maxi]*(amount+1)
        dp[0] = 0
        for c in coins:
            for i in range(c,amount+1):
                dp[i] = min(dp[i],dp[i-c]+1)
                
        if dp[-1]==maxi: return -1
        else: return dp[-1]

File: backend/CDataset/coinChange/coinChange_9.cpp
const int m = 1e4 + 100; // max allowed amount < 1e4
class Solution {
    int dp[m];
public:
    int coinChange(vector<int>& coins, int amount){
        sort(coins.begin(), coins.end()); // sort on basis on coin value
        
        for(int i=0;i<=amount;i++)
            dp[i] = INT_MAX; // initialize
        
        dp[0] = 0; // no coin needed for 0 value
        
        for(int i=coins[0];i<=amount;i++){ // we can only make values > coins[0]
            for(int j=0;j<coins.size();j++){
                if(i < coins[j]) // we cant take any coin which comes after this
                    break;
                if(dp[i-coins[j]]!=INT_MAX) 
                    dp[i] = min(dp[i], 1 + dp[i-coins[j]]);// recurrence obvious
            }
        }
        
        if(dp[amount]==INT_MAX)
            return -1;
        
        return dp[amount];
    }
};
